### PR TITLE
Extend doc keyword feature by allowing any ident

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #51315.

As suggested by @danielhenrymantilla in [this comment](https://github.com/rust-lang/rust/issues/51315#issuecomment-733879934), this PR extends `#[doc(keyword = "...")]` to allow any ident to be used as keyword. The final goal is to allow (proc-)macro crates' owners to write documentation of the keywords they might introduce.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -169,11 +169,19 @@ impl Clean<ExternalCrate> for CrateNum {
                 for attr in attrs.lists(sym::doc) {
                     if let Some(v) = attr.value_str() {
                         if attr.has_name(sym::keyword) {
-                            if v.is_doc_keyword() {
```
Comment (petrochenkov) [inline src/librustdoc/clean/mod.rs:172]: Could you remove `fn is_doc_keyword` as well? This was its only use.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:172]: Sure!

File: src/test/rustdoc-ui/invalid-keyword.stderr [hunk 2]
```diff
@@ -0,0 +1,8 @@
+error: `foo df` is not a valid ident
+  --> $DIR/invalid-keyword.rs:3:7
+   |
+LL | #[doc(keyword = "foo df")]
+   |       ^^^^^^^^^^^^^^^^^^
```
Comment (danielhenrymantilla) [inline src/test/rustdoc-ui/invalid-keyword.stderr:5]: Could this span to the attribute arg exactly?

```suggestion
LL | #[doc(keyword = "foo df")]
   |                 ^^^^^^^^
```

File: src/test/rustdoc-ui/invalid-keyword.stderr [hunk 3]
```diff
@@ -0,0 +1,8 @@
+error: `foo df` is not a valid ident
```
Comment (danielhenrymantilla) [inline src/test/rustdoc-ui/invalid-keyword.stderr:1]:
```suggestion
error: `foo df` is not a valid identifier
```

File: src/librustdoc/clean/mod.rs [hunk 4]
```diff
@@ -169,11 +169,19 @@ impl Clean<ExternalCrate> for CrateNum {
...
-                                keyword = Some(v.to_string());
-                                break;
+                            let k = v.to_string();
+                            if !rustc_lexer::is_ident(&k) {
+                                cx.tcx
+                                    .sess
+                                    .struct_span_err(
+                                        attr.span(),
+                                        &format!("`{}` is not a valid ident", v),
```
Comment (danielhenrymantilla) [inline src/librustdoc/clean/mod.rs:178]: This is how that might be achieved (there might be a better way; I don't know the ast well enough. Outlining that into its own `span_of_value_str` might make it more readable):

```suggestion
                                        attr.meta_item().unwrap().name_value_literal().unwrap().span,
                                        &format!("`{}` is not a valid identifier", v),
```

**Review by danielhenrymantilla — COMMENTED** [review]
Great! Having this will be awesome indeed for proc-macros 😍 

  - Just a minor nit about the span of the error in the non-identifier case.

**Comment by jyn514** [comment]
Rather than removing `is_doc_keyword`, can you add an internal lint that fires if you try to use this on something that's not a keyword? I know there are internal lints for the compiler, not sure if they run on the standard library.

**Comment by GuillaumeGomez** [comment]
You mean for the `std` directly or as a whole?

**Comment by petrochenkov** [comment]
>Rather than removing `is_doc_keyword`, can you add an internal lint that fires if you try to use this on something that's not a keyword? I know there are internal lints for the compiler, not sure if they run on the standard library.

The benefit/cost ratio for this is abysmal, IMO.

File: src/librustdoc/clean/mod.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:178]: Funnily enough, we generally don't indicate the value but the attribute. Might be interesting to improve it in some other places. Noting that for myself haha.

**Comment by GuillaumeGomez** [comment]
Maybe @jyn514 means that we want to add a check for that specifically for the std? If so, I think it's worth it doing it in another PR (but still removing the function here). What do you think?

**Comment by jyn514** [comment]
Yes, I meant for the standard library.

@petrochenkov if you think that's not worth maintaining, I can make a PR doing this once, making sure the lint doesn't trigger, and then just close the PR.

**Review by danielhenrymantilla — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
>  if you think that's not worth maintaining, I can make a PR doing this once, making sure the lint doesn't trigger, and then just close the PR.

@jyn514 Just realized that it's unneeded since it was only allowing official keywords before this PR. Therefore, the `std` literally could not add doc for something else.

**Comment by jyn514** [comment]
@GuillaumeGomez it could, it just wouldn't show up in the documentation. Anyway, I agree it's not worth blocking this PR.

@bors r+

**Pull request merged**